### PR TITLE
OTT-1160: Introduce prometheus metric-engine in openwrap module

### PR DESCRIPTION
**Pull request by ashishshinde-pubm**
Please refer - https://github.com/PubMatic-OpenWrap/prebid-server/pull/526  for review comments

### CODE REVIEW

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+		"inject_tracker_errors",
+		"Count of errors while injecting trackers at publisher, partner level.",
+		[]string{pubIDLabel, partnerLabel, adFormatLabel},
+	)
+
+	metrics.pubPartnerResponseTimeMs = newHistogramVec(cfg, promRegistry,
+		"partner_response_time",
+		"Time taken by each partner to respond in milli-seconds labeled by publisher.",
+		[]string{pubIDLabel, partnerLabel},
+		[]float64{10, 50, 100, 250, 500, 1000},
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:110]: use prebid buckets

**Review by pm-nilesh-chate — APPROVED** [review]
added minor comment

**Comment by pm-nilesh-chate** [comment]
check if `./coverage` script marks this PR green

**Review by pm-nilesh-chate — APPROVED** [review]

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go (hunk 1 above)
Comment (ashishshinde-pubm) [reply modules/pubmatic/openwrap/metrics/prometheus/prometheus.go]: Done

**Review by pm-nilesh-chate — DISMISSED** [review]

**Review by pm-viral-vala — DISMISSED** [review]

File: modules/pubmatic/openwrap/metrics/config/metrics.go [hunk 2]
```diff
@@ -44,21 +48,29 @@ func NewMetricsEngine(cfg config.Config) (MultiMetricsEngine, error) {
 		engineList = append(engineList, sc)
 	}
 
-	// TODO: Set up the Prometheus metrics engine.
+	// Set up the Prometheus metrics engine.
+	if metricsCfg != nil && metricsRegistry != nil && metricsRegistry[metrics_cfg.PrometheusRegistry] != nil {
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply modules/pubmatic/openwrap/metrics/config/metrics.go:52]: The Prometheus registry is initialised in router.go and then set under the metricsRegistry. 
Hence in NewMetricsEngine function, we will only read it. If registry is absent then we will not initialise Prometheus metric-engine.

File: metrics/config/metrics_test.go [hunk 3]
```diff
@@ -31,13 +32,85 @@ func TestGoMetricsEngine(t *testing.T) {
...
 	syncerKeys := []string{"keyA", "keyB"}
-	testEngine := NewMetricsEngine(&cfg, adapterList, syncerKeys, modulesStages)
+	testEngine := NewMetricsEngine(&cfg, NewMetricsRegistry(), adapterList, syncerKeys, modulesStages)
 	_, ok := testEngine.MetricsEngine.(*metrics.Metrics)
 	if !ok {
 		t.Error("Expected a Metrics as MetricsEngine, but didn't get it")
 	}
 }
 
+func TestGoMetricsEngineForNilRegistry(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/config/metrics_test.go:42]: Done.

File: metrics/config/metrics.go [hunk 4]
```diff
@@ -3,17 +3,35 @@ package config
...
+)
+
+// NewMetricsRegistry returns the map of metrics-engine-name and its respective registry
+func NewMetricsRegistry() MetricsRegistry {
+	return MetricsRegistry{
+		PrometheusRegistry: prometheus.NewRegistry(),
+		InfluxRegistry:     gometrics.NewPrefixedRegistry("prebidserver."),
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/config/metrics.go:31]: Done. this will minimise the prebid merge conflict during upgrade

**Review by pm-nilesh-chate — APPROVED** [review]

**Review by pm-saurabh-narkhede — APPROVED** [review]

**Review by pm-viral-vala — APPROVED** [review]

**Pull request merged**